File: problems/easy/0020-valid-parentheses/python/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        opening_brackets = {"(", "[", "{"}

        pair = {
            ")": "(",
            "]": "[",
            "}": "{",
        }

        for char in s:
            if char in opening_brackets:
                stack.append(char)

            else:
                if len(stack) == 0: 
                    return False

                top = stack[-1]

                if top == pair[char]:
                    stack.pop()

                else:
                    return False

        return len(stack) == 0

File: problems/easy/0020-valid-parentheses/typescript/solution.ts
export function isValid(s: string): boolean {
    const stack: string[] = [];
    const pairs: { [key: string]: string } = {
        ')': '(',
        ']': '[',
        '}': '{'
    };

    for(let i = 0; i < s.length; i++){
        const char = s[i];

        if(char === '(' || char === '[' || char === '{'){
            stack.push(char);
        } 
        
        else {
            if (stack.length === 0){
                return false;
            }

            const top = stack[stack.length - 1];

            if(top === pairs[char]){
                stack.pop();
            } else {
                return false;
            }
        }
    }

    return stack.length === 0;
};